satellite/gc/piecetracker.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gc
+
+import (
+	"context"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	monkit "gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/pkg/bloomfilter"
+	"storj.io/storj/pkg/overlay"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+var (
+	mon = monkit.Package()
+
+	// Error defines the piece tracker errors class
+	Error = errs.Class("piece tracker error")
+)
+
+// RetainInfo contains info needed for a storage node to retain important data and delete garbage data
+type RetainInfo struct {
+	Filter       *bloomfilter.Filter
+	CreationDate time.Time
+	address      *pb.NodeAddress
+	count        int
+}
+
+// PieceTracker allows access to info about the good pieces that storage nodes need to retain
+type PieceTracker interface {
+	// Add adds a RetainInfo to the PieceTracker
+	Add(ctx context.Context, nodeID storj.NodeID, pieceID storj.PieceID) error
+	// GetRetainInfos gets all of the RetainInfos
+	GetRetainInfos() map[storj.NodeID]*RetainInfo
+}
+
+// pieceTracker contains info about the good pieces that storage nodes need to retain
+type pieceTracker struct {
+	log                *zap.Logger
+	overlay            overlay.DB
+	filterCreationDate time.Time
+	initialPieces      int64
+	falsePositiveRate  float64
+	retainInfos        map[storj.NodeID]*RetainInfo
+	pieceCounts        map[storj.NodeID]int
+}
+
+// Add adds a pieceID to the relevant node's RetainInfo
+func (pieceTracker *pieceTracker) Add(ctx context.Context, nodeID storj.NodeID, pieceID storj.PieceID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	var filter *bloomfilter.Filter
+
+	// If we know how many pieces a node should be storing, use that number. Otherwise use default.
+	numPieces := int(pieceTracker.initialPieces)
```

Comment:
it's not also needed below in this check?
```
	if pieceTracker.pieceCounts[nodeID] > 0 {
		numPieces = pieceTracker.pieceCounts[nodeID]
	}
```

Reply:
I think Ivan might mean taking lines 60 through 64, and moving them all inside the `if` statement that starts at line 65. Because if `ok` is true on line 65, then `numPieces` isn't needed at all, and we can save a microsecond or two by not doing map lookups on `pieceTracker`.

But I don't see any real harm to doing it the way you've got it.

Reply:
ah okay, I'll just go ahead and change it.

Reply:
Not harm and I didn't comment for the sake of the minor performance grain that it may produce (which is probably negligible and without profiling I would never suggest that change), mostly to keep those variables in the scope that they are used which as a reader of the code allows me to not have to retain them in memory after outside of the scope that they are used when I read the code.

Probably, I comment these things to survive with my :fish: brain